demos/python_demos/formula_recognition_demo/formula_recognition_demo.py

```diff
@@ -16,68 +13,131 @@
-
-
-def preprocess_image(preprocess, image_raw, tgt_shape):
-    target_height, target_width = tgt_shape
-    image_raw = preprocess(image_raw, tgt_shape)
-    img_h, img_w = image_raw.shape[0:2]
-    image_raw = cv.copyMakeBorder(image_raw, 0, target_height - img_h,
-                                  0, target_width - img_w, cv.BORDER_CONSTANT,
-                                  None, COLOR_WHITE)
-    image = image_raw.transpose((2, 0, 1))
-    return np.expand_dims(image, axis=0)
+from utils import *
+
+CONFIDENCE_THRESH = 0.95
+
+
+class InteractiveDemo:
+    def __init__(self, input_model_shape, resolution=DEFAULT_RESOLUTION):
+        self.resolution = resolution
+        self._tgt_shape = input_model_shape
+        self.start_point, self.end_point = self._create_input_window()
+        self._prev_rendered_formula = None
+        self._prev_formula_img = None
+        self._latex_h = 0
+        self._renderer = create_renderer()
+
+    def _create_input_window(self):
+        aspect_ratio = self._tgt_shape[0] / self._tgt_shape[1]
+        width = min(DEFAULT_WIDTH, self.resolution[0])
+        height = int(width * aspect_ratio)
+        start_point = (int(self.resolution[0] / 2 - width / 2), int(self.resolution[1] / 2 - height / 2))
+        end_point = (int(self.resolution[0] / 2 + width / 2), int(self.resolution[1] / 2 + height / 2))
+        return start_point, end_point
+
+    def get_crop(self, frame):
+        crop = frame[self.start_point[1]:self.end_point[1], self.start_point[0]:self.end_point[0], :]
+        return crop
+
+    def _draw_rectangle(self, frame):
+        frame = cv.rectangle(frame, self.start_point, self.end_point, color=RED, thickness=2)
+        return frame
+
+    def resize_window(self, action):
+        height = self.end_point[1] - self.start_point[1]
+        width = self.end_point[0] - self.start_point[0]
+        aspect_ratio = height / width
+        if action == 'increase':
+            max_h, max_w = min(MAX_HEIGHT, self.resolution[1]), min(MAX_WIDTH, self.resolution[0])
+            if height >= max_h or width >= max_w:
+                return
+            self.start_point = (self.start_point[0]-DEFAULT_RESIZE_STEP,
+                                self.start_point[1] - int(DEFAULT_RESIZE_STEP * aspect_ratio))
+            self.end_point = (self.end_point[0]+DEFAULT_RESIZE_STEP,
+                              self.end_point[1] + int(DEFAULT_RESIZE_STEP * aspect_ratio))
+        elif action == 'decrease':
+            if height <= MIN_HEIGHT or width <= MIN_WIDTH:
+                return
+            self.start_point = (self.start_point[0]+DEFAULT_RESIZE_STEP,
+                                self.start_point[1] + int(DEFAULT_RESIZE_STEP * aspect_ratio))
+            self.end_point = (self.end_point[0]-DEFAULT_RESIZE_STEP,
+                              self.end_point[1] - int(DEFAULT_RESIZE_STEP * aspect_ratio))
+        else:
+            raise ValueError(f"wrong action: {action}")
```

Comment:
I'm not sure we dropped old python versions support which can't handle such formatting.

Reply:
Changed f-strings to `.format` method